### PR TITLE
React docs list

### PR DESCRIPTION
**Pull request by garrensmith**
This is the document list for views, _all_docs redone in React.js

### CODE REVIEW

File: app/addons/documents/index-results/tests/index-results.actionsSpec.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
+define([
```
Comment (kxepal) [inline app/addons/documents/index-results/tests/index-results.actionsSpec.js]: License header is missed

File: app/addons/components/assets/less/docs.less [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    label {
+      padding-left: 23px; // 7px to the right-border + 16px around
+    }
+  }
+  .doc-item {
+    width: auto;
+    overflow: hidden;
+  }
+  label.label-checkbox-doclist {
+    // https://code.google.com/p/chromium/issues/detail?id=411065
```
Comment (kxepal) [inline app/addons/components/assets/less/docs.less]: This bug looks as fixed in November of the past year. Still actual?

File: app/addons/documents/index-results/actions.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      });
+
+      //show menu
+      if (!headerBarStore.getToggleStatus()) {
+        HeaderActions.toggleHeaderControls();
+        return;
+      }
+
+      //hide menu
+      if(headerBarStore.getToggleStatus && indexResultsStore.getSelectedItemsLength() === 0) {
```
Comment (kxepal) [inline app/addons/documents/index-results/actions.js]: Shouldn't here be `headerBarStore.getToggleStatus()`?

File: app/addons/documents/header/header.actiontypes.js [hunk 4]
```diff
@@ -14,9 +14,11 @@ define([], function () {
   return {
     UPDATE_DOCUMENT_COUNT: 'UPDATE_DOCUMENT_COUNT',
     COLLAPSE_DOCUMENTS: 'COLLAPSE_DOCUMENTS',
+    UNCOLLAPSE_DOCUMENTS: 'UNCOLLAPSE_DOCUMENTS',
```
Comment (kxepal) [inline app/addons/documents/header/header.actiontypes.js]: Isn't Expand an opposite word for Collapse?

File: app/addons/documents/tests/resourcesSpec.js [hunk 5]
```diff
@@ -233,28 +238,61 @@ define([
...
+        {"error":"conflict","id":"Deferred","rev":"10-72cd2edbcc0d197ce96188a229a7af01"},
         {"ok":true,"id":"DeskSet","rev":"6-da537822b9672a4b2f42adb1be04a5b1"}
-      ]);
+      ], promise);
     });
 
     it("removes successfull deleted from the collection but keeps one with errors", function () {
+      var promise = FauxtonAPI.Deferred();
       collection.handleResponse([
         {"error":"confclict","id":"1","rev":"10-72cd2edbcc0d197ce96188a229a7af01"},
```
Comment (kxepal) [inline app/addons/documents/tests/resourcesSpec.js]: You'd fixed that typo above, but not here.

File: app/addons/documents/tests/resourcesSpec.js [hunk 6]
```diff
@@ -233,28 +238,61 @@ define([
...
+    });
+
+    it('resolves promise with list of errored ids', function () {
+      var promise = FauxtonAPI.Deferred();
+      var spy = sinon.spy();
+
+      promise.then(spy);
+
+      collection.handleResponse([
+        {"error":"confclict","id":"1","rev":"10-72cd2edbcc0d197ce96188a229a7af01"},
```
Comment (kxepal) [inline app/addons/documents/tests/resourcesSpec.js]: And copy-pasted it here (:

File: app/addons/documents/header/header.actiontypes.js (hunk 4 above)
Comment (garrensmith) [inline app/addons/documents/header/header.actiontypes.js]: Ha, good spot.

**Comment by michellephung** [comment]
Hey garren,

I can't seem to select a particular document using the gray circle next to it: 
screen shot 2015-03-20 at 10 34 03 am

the header above will flash, but then the document is not selected. 
No console errors shown.

File: app/addons/documents/tests/nightwatch/viewQueryOptions.js [hunk 7]
```diff
@@ -13,7 +13,7 @@
 module.exports = {
   'Edit view: Queryoptions work': function (client) {
     /*jshint multistr: true */
-    var waitTime = 10000,
+    var waitTime = 50000,
```
Comment (robertkowalski) [inline app/addons/documents/tests/nightwatch/viewQueryOptions.js]: i think we can revert that

File: app/addons/documents/pagination/actions.js [hunk 8]
```diff
@@ -67,10 +62,14 @@ function (app, FauxtonAPI, ActionTypes, Stores) {
         type: ActionTypes.PAGINATE_PREVIOUS,
       });
 
-      FauxtonAPI.triggerRouteEvent('paginate', {
+      /*FauxtonAPI.triggerRouteEvent('paginate', {
        direction: 'previous',
        perPage: store.getPerPage(),
        currentPage: store.getCurrentPage()
+     });*/
```
Comment (robertkowalski) [inline app/addons/documents/pagination/actions.js]: can we delete it?

File: app/addons/documents/pagination/actions.js [hunk 9]
```diff
@@ -44,21 +41,19 @@ function (app, FauxtonAPI, ActionTypes, Stores) {
...
       FauxtonAPI.dispatch({
         type: ActionTypes.PAGINATE_NEXT,
       });
 
-      FauxtonAPI.triggerRouteEvent('paginate', {
+      /*FauxtonAPI.triggerRouteEvent('paginate', {
        direction: 'next',
        perPage: store.documentsLeftToFetch(),
        currentPage: store.getCurrentPage()
+     });*/
```
Comment (robertkowalski) [inline app/addons/documents/pagination/actions.js]: can we delete it?

File: app/addons/documents/index-results/tests/index-results.storesSpec.js [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+        assert.ok(store.canUncollapseDocs());
+    });
+
+    it('returns false for no collapsed docs', function () {
+        store.clearCollapsedDocs();
+
+        assert.notOk(store.canUncollapseDocs());
+    });
+
+  });
```
Comment (robertkowalski) [inline app/addons/documents/index-results/tests/index-results.storesSpec.js]: from the beginning of the file until here 4 spaces are uses as intendation

File: app/addons/documents/views.js [hunk 11]
```diff
@@ -269,7 +269,7 @@ function (app, FauxtonAPI, Components, Documents,
     }
   });
 
-  Views.AllDocsList = FauxtonAPI.View.extend({
+  /*Views.AllDocsList = FauxtonAPI.View.extend({
```
Comment (robertkowalski) [inline app/addons/documents/views.js]: we can delete that - yay!!! :)

File: app/addons/documents/header/header.actions.js [hunk 12]
```diff
@@ -37,13 +51,13 @@ function (app, FauxtonAPI, ActionTypes) {
       });
     },
 
-    deleteSelected: function () {
+    /*deleteSelected: function () {
```
Comment (robertkowalski) [inline app/addons/documents/header/header.actions.js]: can we delete that?

**Comment by robertkowalski** [comment]
tests are failing

File: app/addons/documents/header/header.actiontypes.js [hunk 13]
```diff
@@ -14,9 +14,11 @@ define([], function () {
   return {
     UPDATE_DOCUMENT_COUNT: 'UPDATE_DOCUMENT_COUNT',
     COLLAPSE_DOCUMENTS: 'COLLAPSE_DOCUMENTS',
+    EXPAND_DOCUMENTS: 'UNCOLLAPSE_DOCUMENTS',
     TOGGLE_HEADER_CONTROLS: 'TOGGLE_HEADER_CONTROLS',
     RESET_HEADER_BAR: 'RESET_HEADER_BAR',
-    DELETE_SELECTED_DOCUMENTS: 'DELETE_SELECTED_DOCUMENTS'
+    DELETE_SELECTED_DOCUMENTS: 'DELETE_SELECTED_DOCUMENTS',
```
Comment (robertkowalski) [inline app/addons/documents/header/header.actiontypes.js]: seems this is not used any more

File: app/addons/documents/index-results/stores.js [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    isDeleteable: function () {
+      return this._deleteable;
+    },
+
+    selectDoc: function (id) {
+      if (!this._selectedItems[id]) {
+        this._selectedItems[id] = true;
+      } else {
+        delete this._selectedItems[id];
+      }
```
Comment (robertkowalski) [inline app/addons/documents/index-results/stores.js]: you can use an early return here and get rid of the else

File: app/addons/documents/tests/actionsSpec.js [hunk 15]
```diff
@@ -46,8 +48,9 @@ define([
       });
 
       afterEach(function () {
-        FauxtonAPI.navigate.restore && FauxtonAPI.navigate.restore();
-        FauxtonAPI.triggerRouteEvent.restore && FauxtonAPI.triggerRouteEvent.restore();
+        restore(FauxtonAPI.navigate);
+        restore(FauxtonAPI.triggerRouteEvent);
+        restore(IndexResultsActions.reloadResultsList);
```
Comment (robertkowalski) [inline app/addons/documents/tests/actionsSpec.js]: cool!

File: app/addons/documents/index-results/index-results.components.react.jsx [hunk 16]
```diff
@@ -28,10 +36,118 @@ function (app, FauxtonAPI, React) {
...
+      var classNames = 'view';
+
+      if (this.props.isDeleteable) {
+        classNames += ' show-select';
+      }
+
+      return (
+      <div className={classNames}>
+        <div id="doc-list">
+          <ReactCSSTransitionGroup transitionName={'slow-fade'}>
```
Comment (robertkowalski) [inline app/addons/documents/index-results/index-results.components.react.jsx]: can we make that a very fast fade? maybe 500ms

it irritates a bit when i want to go two pages forward

Comment (robertkowalski) [inline app/addons/documents/index-results/index-results.components.react.jsx]: maybe even just 250 or 300

File: app/addons/documents/routes-documents.js [hunk 17]
```diff
@@ -119,7 +116,7 @@ function(app, FauxtonAPI, BaseRoute, Documents, Changes, Index, DocEditor,
       this.rightHeader.hideQueryOptions();
 
       this.apiUrl = [designDocInfo.url('apiurl'), designDocInfo.documentation()];
-    },
+      },
```
Comment (robertkowalski) [inline app/addons/documents/routes-documents.js]: whoops!

**Comment by michellephung** [comment]
+1

File: app/addons/documents/tests/nightwatch/paginateView.js [hunk 18]
```diff
@@ -27,11 +27,12 @@ module.exports = {
       // hack to get select working by clicking on it and using keyboard to select
       // http://www.w3.org/TR/2012/WD-webdriver-20120710/
       .keys(['\uE013', '\uE006'])
-      .waitForElementNotPresent('.spinner', waitTime)
+      .waitForElementNotPresent('div[data-id="document_9"]', waitTime)
       .execute(function () {
         return $('.doc-row').length;
       }, function (result) {
-          client.assert.equal(result.value, 10);
+        console.log(result.value);
```
Comment (benkeen) [inline app/addons/documents/tests/nightwatch/paginateView.js]: Debugging code

File: app/addons/documents/index-results/stores.js [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+      var doc = originalDoc.toJSON();
+
+      if (this.isCollapsed(doc._id)) {
+        doc = {
+          _id: _.isUndefined(doc._id),
+          _rev: doc._rev
+        };
+      }
+
+      return JSON.stringify(doc, null, "  ");
```
Comment (robertkowalski) [inline app/addons/documents/index-results/stores.js]: single quotes

File: app/addons/documents/header/header.actions.js [hunk 20]
```diff
@@ -18,31 +18,37 @@ define([
...
 
-      FauxtonAPI.Events.trigger('headerbar:collapse');
     },
 
-    toggleSelectAllDocuments: function (on) {
-      FauxtonAPI.Events.trigger('headerbar:selectall', on);
+    unCollapseDocuments: function () {
+      FauxtonAPI.dispatch({
+        type: ActionTypes.EXPAND_DOCUMENTS
+      });
```
Comment (robertkowalski) [inline app/addons/documents/header/header.actions.js]: 2 spaces intendation

File: app/addons/documents/header/header.react.jsx [hunk 21]
```diff
@@ -39,11 +43,11 @@ function (app, FauxtonAPI, React, Stores, Actions, ReactComponents) {
     },
 
     componentDidMount: function () {
-      bulkDocumentHeaderStore.on('change', this.onChange, this);
+      indexResultsStore.on('change', this.onChange, this);
     },
 
     componentWillUnmount: function() {
```
Comment (benkeen) [inline app/addons/documents/header/header.react.jsx]: space after function

File: app/addons/documents/index-results/actions.js [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+      });
+
+      //show menu
+      if (!headerBarStore.getToggleStatus()) {
+        HeaderActions.toggleHeaderControls();
+        return;
+      }
+
+      //hide menu
+      if(headerBarStore.getToggleStatus() && indexResultsStore.getSelectedItemsLength() === 0) {
```
Comment (benkeen) [inline app/addons/documents/index-results/actions.js]: space after if

File: app/addons/documents/index-results/index-results.components.react.jsx [hunk 23]
```diff
@@ -13,10 +13,18 @@
...
   'app',
   'api',
-  'react'
+  'react',
+  'addons/documents/index-results/stores',
+  'addons/documents/index-results/actions',
+  'addons/components/react-components.react',
+  'addons/documents/resources',
+
+  "plugins/prettify"
```
Comment (benkeen) [inline app/addons/documents/index-results/index-results.components.react.jsx]: single quotes

File: app/addons/documents/index-results/index-results.components.react.jsx [hunk 24]
```diff
@@ -28,10 +36,118 @@ function (app, FauxtonAPI, React) {
...
+           onDoubleClick={this.onDoubleClick}
+           keylabel={doc.keylabel}
+           docContent={doc.content}
+           checked={this.props.isSelected(doc.id)}
+           docChecked={this.props.docChecked}
+           docIdentifier={doc.id} >
+           {this.getUrlFragment('#' + doc.url)}
+         </Components.Document>
+       );
+     }, this);
```
Comment (benkeen) [inline app/addons/documents/index-results/index-results.components.react.jsx]: spacing

File: app/addons/documents/index-results/index-results.components.react.jsx [hunk 25]
```diff
@@ -28,10 +36,118 @@ function (app, FauxtonAPI, React) {
...
+
+    isSelected: function (id) {
+      return this.state.isSelected(id);
+    },
+
+    getInitialState: function () {
+      return this.getStoreState();
+    },
+
+    componentDidMount: function() {
```
Comment (benkeen) [inline app/addons/documents/index-results/index-results.components.react.jsx]: space after function, same for next function

File: app/addons/documents/index-results/stores.js [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+define([
+  'api',
+  'addons/documents/index-results/actiontypes',
+  'addons/documents/header/header.actiontypes',
+  "addons/documents/resources"
```
Comment (benkeen) [inline app/addons/documents/index-results/stores.js]: single quotes

File: app/addons/documents/index-results/stores.js [hunk 27]
```diff
@@ -0,0 +1,270 @@
...
+],
+
+function(FauxtonAPI, ActionTypes, HeaderActionTypes, Documents) {
```
Comment (benkeen) [inline app/addons/documents/index-results/stores.js]: space after function

**Comment by benkeen** [comment]
This looks & functions great, @garrensmith! Really like the fade effect on pagination - much smoother than before. 

Collapse/Expand also works well and I like that it remembers the checked state when you go between pages. If you don't want to change it to fully collapse the docs to a single row with this ticket, mind if I tackle it separately in a follow-up ticket?

**Comment by robertkowalski** [comment]
please rebase against master and run: 

```
grunt couchdb && npm i && npm run stylecheck
```

**Comment by garrensmith** [comment]
@robertkowalski and @benkeen I've fixed all the style issues.

@benkeen you can definitely implement the full collapse in another PR once this lands.

**Comment by robertkowalski** [comment]
+1

**Comment by garrensmith** [comment]
Yes! Finally. All tests passing [here](https://travis-ci.org/garrensmith/couchdb-fauxton/builds/55928223)

**Pull request merged**